### PR TITLE
Yubico ads conversion script

### PR DESCRIPTION
**Pull request by amirsaber**

### CODE REVIEW

**Comment by amirsaber** [comment]
Check for enabled ads is missing

File: Greaselion.json [hunk 1]
```diff
@@ -204,5 +204,16 @@
     "preconditions": {
       "ads-enabled": true
     }
+  },
+  {
+    "urls": [
+      "https://*yubico.com*/checkout/confirmation*"
```
Comment (emerick) [inline Greaselion.json:210]: The stars in the host portion of this url seem problematic. Wouldn't that end up running this script on potentially unrelated sites?

File: scripts/brave_rewards/ads/conversions/yubico.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const initYubicoScript = () => {
+  const el = document.getElementsByClassName('order-info')
+  let orderId = ''
+  if (el && el[0] && el[0].firstChild && el[0].firstChild.firstChild) {
+    orderId = el[0].firstChild.firstChild.textContent!.replace('Order Number: ', '')
```
Comment (emerick) [inline scripts/brave_rewards/ads/conversions/yubico.ts:9]: Is the `!` a typo?

File: scripts/brave_rewards/ads/conversions/yubico.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  if (orderId !== '') {
+    const link = document.createElement('meta')
+    link.setAttribute('name', 'ad-conversion-id')
+    link.content = orderId
+    document.getElementsByTagName('head')[0].appendChild(link)
```
Comment (emerick) [inline scripts/brave_rewards/ads/conversions/yubico.ts:16]: Probably would be good to do some minimal null checks here.

File: scripts/brave_rewards/ads/conversions/yubico.ts (hunk 2 above)
Comment (amirsaber) [reply scripts/brave_rewards/ads/conversions/yubico.ts:9]: no it's not, it's a Non-null assertion operator

File: scripts/brave_rewards/ads/conversions/yubico.ts (hunk 3 above)
Comment (amirsaber) [reply scripts/brave_rewards/ads/conversions/yubico.ts:16]: what line could possibly go null?

File: scripts/brave_rewards/ads/conversions/yubico.ts (hunk 2 above)
Comment (emerick) [reply scripts/brave_rewards/ads/conversions/yubico.ts:9]: Didn't realize that, nice!

File: scripts/brave_rewards/ads/conversions/yubico.ts (hunk 3 above)
Comment (emerick) [reply scripts/brave_rewards/ads/conversions/yubico.ts:16]: Is `getElementsByTagName('head')` guaranteed to have an element? I think that's the only one I was concerned about, but maybe it's fine.

File: Greaselion.json (hunk 1 above)
Comment (amirsaber) [reply Greaselion.json]: pushed a change to target domain and subdomains explicitly

Comment (emerick) [reply Greaselion.json]: Is the `.com*` intentional? It doesn't seem like you need a `*` there, but could be I just don't understand the goal.

File: Greaselion.json [hunk 4]
```diff
@@ -204,5 +204,17 @@
...
+  {
+    "urls": [
+      "https://*.yubico.com*/checkout/confirmation*",
+      "https://yubico.com*/checkout/confirmation*"
+    ],
+    "scripts": [
+      "scripts/brave_rewards/ads/conversions/yubico.bundle.js"
+    ],
+    "preconditions": {
+      "ads-enabled": true
```
Comment (diracdeltas) [inline Greaselion.json:217]: @emerick did we ever add the feature to disable greaselion scripts in incognito by default?

File: Greaselion.json [hunk 5]
```diff
@@ -204,5 +204,17 @@
     "preconditions": {
       "ads-enabled": true
     }
+  },
+  {
+    "urls": [
+      "https://*.yubico.com*/checkout/confirmation*",
+      "https://yubico.com*/checkout/confirmation*"
```
Comment (diracdeltas) [inline Greaselion.json:211]: what's the code that parses these match patterns? would these match `https://yubico.com.example.com` for instance?

File: scripts/brave_rewards/ads/conversions/yubico.ts (hunk 2 above)
Comment (diracdeltas) [inline scripts/brave_rewards/ads/conversions/yubico.ts:9]: will this work for non-English versions of the page? (assuming they have localization)

File: webpack.config.js [hunk 6]
```diff
@@ -28,6 +28,7 @@ module.exports = (env, argv) => {
       ['scripts/brave_rewards/publisher/vimeo/vimeoAutoContribution']: './scripts/brave_rewards/publisher/vimeo/vimeoAutoContribution',
       ['scripts/brave_rewards/publisher/youtube/youtubeBase']: './scripts/brave_rewards/publisher/youtube/youtubeBase',
       ['scripts/brave_rewards/publisher/youtube/youtubeAutoContribution']: './scripts/brave_rewards/publisher/youtube/youtubeAutoContribution',
+      ['scripts/brave_rewards/ads/conversions/yubico']: './scripts/brave_rewards/ads/conversions/yubico',
```
Comment (diracdeltas) [inline webpack.config.js:31]: should have a wiki page to document why this script is injected. we already have users complaining that the greaselion injected scripts are not transparent enough and can't be disabled.

File: Greaselion.json (hunk 4 above)
Comment (emerick) [reply Greaselion.json:217]: It hasn't been implemented yet (https://github.com/brave/brave-browser/issues/11409), I'll bring it up at our triage meeting this week!

**Review by orspetol — CHANGES_REQUESTED** [review]

File: webpack.config.js (hunk 6 above)
Comment (amirsaber) [reply webpack.config.js:31]: I have a wiki update page written down in the spec, awaiting for the final approval and copy edits on it

File: scripts/brave_rewards/ads/conversions/yubico.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const initYubicoScript = () => {
+  const el = document.getElementsByClassName('order-info')
+  let orderId = ''
+  if (el && el[0] && el[0].firstChild && el[0].firstChild.firstChild) {
+    orderId = el[0].firstChild.firstChild.textContent!.replace('Order Number: ', '')
+  }
+
+  if (orderId !== '') {
```
[the remark this answers is not in the document]
Comment (amirsaber) [reply scripts/brave_rewards/ads/conversions/yubico.ts:13]: added substr call on orderId in next line

**Review by diracdeltas — CHANGES_REQUESTED** [review]
We have an ongoing relationship with Yubico, so let's first ask them if they can make this change without Brave having to inject scripts.

File: Greaselion.json (hunk 1 above)
Comment (amirsaber) [reply Greaselion.json]:
```
      "https://*.yubico.com/*/checkout/confirmation*",
      "https://yubico.com/*/checkout/confirmation*"
```
 moved the asterisk to after domain separation from address

File: Greaselion.json (hunk 5 above)
Comment (amirsaber) [reply Greaselion.json]: also mentioned by @emerick in above comments, moved asterisk to after `/`

File: Greaselion.json (hunk 4 above)
Comment (diracdeltas) [reply Greaselion.json:217]: @emerick will a script with `ads-enabled` true get injected into incognito/tor windows?

Comment (emerick) [reply Greaselion.json:217]: I believe so, as it seems that supporting incognito mode is the default for extensions.

Comment (emerick) [reply Greaselion.json:217]: Actually, I didn't realize that we've recently made the following change to thje Greaselion service to disallow incognito injection: https://github.com/brave/brave-core/commit/6bbdfb1137381be8787bdb39462c544446f8ce54. So I think we should be good on that front.

**Comment by diracdeltas** [comment]
Yubico has agreed to work on this from their end: https://bravesoftware.slack.com/archives/CJMAD25L7/p1616449958005200?thread_ts=1616446792.001900&cid=CJMAD25L7. Can we close this out for now?

**Comment by amirsaber** [comment]
Closing this in favor of better approach

**Pull request closed without merging**